### PR TITLE
Modify aggressive fusion heuristics to not create ops where reduction ops also return intermediate results.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Currently under aggressive fusion, elementwise parallel ops are fused with consumer that are reductions. When the parallel op has other uses the fused op also returns the reduced and non-reduced values. This results in operations that are reductions returning results that have shape similar to the entire iteration space of the op. Such ops are hard to handle in the backend.  This patch avoids creating such ops, but adapts the fusion heuristics to still pull these ops into the same dispatch.

### CODE REVIEW

**Comment by pzread** [comment]
Number of dispatches with aggressive fusion after this patch
```
| Model                 | Dispatches |
|-----------------------|------------|
| BertForMaskedLMTF     | 187        |
| DeepLabV3_fp32        | 78         |
| EfficientNet_int8     | 102        |
| EfficientNetV2STF     | 276        |
| MiniLML12H384Uncased  | 184        |
| MobileBertSquad_fp16  | 727        |
| MobileBertSquad_fp32  | 703        |
| MobileBertSquad_int8  | 1403       |
| MobileNetV1_fp32      | 57         |
| MobileNetV2_fp32      | 73         |
| MobileNetV3Small_fp32 | 96         |
| MobileSSD_fp32        | 121        |
| PersonDetect_int8     | 58         |
| PoseNet_fp32          | 48         |
| Resnet50TF            | 76         |
```

**Comment by MaheshRavishankar** [comment]
> Number of dispatches with aggressive fusion after this patch
> 
> Model	Dispatches
> BertForMaskedLMTF	187
> DeepLabV3_fp32	78
> EfficientNet_int8	102
> EfficientNetV2STF	276
> MiniLML12H384Uncased	184
> MobileBertSquad_fp16	727
> MobileBertSquad_fp32	703
> MobileBertSquad_int8	1403
> MobileNetV1_fp32	57
> MobileNetV2_fp32	73
> MobileNetV3Small_fp32	96
> MobileSSD_fp32	121
> PersonDetect_int8	58
> PoseNet_fp32	48
> Resnet50TF	76

Awesome! I was just going to do this!

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp [hunk 1]
```diff
@@ -359,10 +391,8 @@ static bool hasCompatibleOuterParallelLoops(
   llvm::SmallBitVector consumerParallelLoops =
       getOuterParallelLoops(cast<TilingInterface>(consumer.getOperation()));
 
-  if (allowConsumerParallelismPessimization) {
-    if (producerParallelLoops.count() > consumerParallelLoops.count())
```
Comment (pzread) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:363]: Do we want to fuse even if `producerParallelLoops.count() > consumerParallelLoops.count()`? This condition seems to be removed.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp [hunk 2]
```diff
@@ -375,12 +405,12 @@ static bool hasCompatibleOuterParallelLoops(
   }
 
   /// Project out the non-parallel dimensions.
-  llvm::SmallBitVector producerProjectedDims(producerParallelLoops);
+  llvm::SmallBitVector producerProjectedDims(rootOuterParallelLoops);
```
Comment (pzread) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:408]: In the case of fusing with the producer, `rootOuterParallelLoops` will be the consumer parallel loop? Is it fine to always use `rootOuterParallelLoops` here?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp [hunk 3]
```diff
@@ -556,8 +593,7 @@ static bool isFusableWithProducer(OpOperand &operand, bool aggressiveFusion) {
     return false;
   }
 
-  return areOpsAggresiveFusable(producer, consumer,
-                                /*allowConsumerParallelismPessimization=*/false,
```
Comment (pzread) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:560]: I didn't remember the reason we disabled consumer parallelism pessimization, do we want to drop that constraint here?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:363]: The overall logic is changed as follows
1) Now when we find the "root operation" of the dispatch we record the outer parallel loop structure of the op. Then for each op fused into the dispatch, we check if the outer parallel loop structure of this op matches the root op.
2) The "match" is done by
   (a) if the pulled in op is fully parallel, as long as it has many parallel loops as the root op, it is pulled in
   (b) if the pulled in op is not fully parallel (has reductions), then the outer parallel loop structure has to match exactly the outer parallel loop structure of the root op.

Impact is earlier we would only fuse consumer ops that are all parallel but might have more parallel loops than the root. Now we do that with the producer too.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:408]: Yes, we are checking here that just taking the outer parallel loops of the root op (we have checked that the outer parallel loops of producer and consumer match the root op) if the indexing map is indentity on both .Thats sufficient condition to know that both the producer and the consumer have compatible outer parallel loops with the root op (transitively).

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:560]: That was being conservative. We are dropping that now. Also the `allowConsumerParallelismPessimization` was a bit of a misnomer... now the same logic is used to check for "fusability" whether we are fusing with producer or consumer.

**Review by pzread — APPROVED** [review]

**Pull request merged**